addons/webhook/certificates.go

```diff
@@ -0,0 +1,203 @@
+		Subject: pkix.Name{
+			CommonName: fmt.Sprintf("%s@%d", tls.Hostname, time.Now().Unix()),
+		},
+		NotBefore: tls.ValidFrom,
+		NotAfter:  tls.ValidFrom.Add(tls.MaxAge),
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+	cacert, err := x509.ParseCertificate(*tls.ca_cert_der)
+	if err != nil {
+		return nil, err
+	}
+	cert, err = x509.CreateCertificate(cryptorand.Reader, &serverTemlate, cacert, &tls.server_key_der.PublicKey, tls.ca_key_der)
+	if err != nil {
+		return nil, err
+	}
+	tls.server_cert_der = &cert
+	return &tls, nil
+}
+
+func (tls *TLSCredentials) WriteFiles() error {
+	//both ca and server certificates
+	certBuffer := bytes.Buffer{}
+	if err := pem.Encode(&certBuffer, &pem.Block{Type: CertificateBlockType, Bytes: *tls.server_cert_der}); err != nil {
+		return err
+	}
+	if err := pem.Encode(&certBuffer, &pem.Block{Type: CertificateBlockType, Bytes: *tls.ca_cert_der}); err != nil {
+		return err
+	}
+	// Server key
+	keyBuffer := bytes.Buffer{}
+	if err := pem.Encode(&keyBuffer, &pem.Block{Type: keyutil.RSAPrivateKeyBlockType, Bytes: x509.MarshalPKCS1PrivateKey(tls.server_key_der)}); err != nil {
+		return err
+	}
+
+	if err := certutil.WriteCert(certfile_path(), certBuffer.Bytes()); err != nil {
+		return err
+	}
+	if err := keyutil.WriteKey(keyfile_path(), keyBuffer.Bytes()); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// from controller-rutime documentation at https://github.com/kubernetes-sigs/controller-runtime/blob/3dafc3c844685e4f672d9e07b88b68ae3f14a842/pkg/config/v1alpha1/types.go#L141
+// CertDir is the directory that contains the server key and certificate.
+// if not set, webhook server would look up the server key and certificate in
+// {TempDir}/k8s-webhook-server/serving-certs. The server key and certificate
+// must be named tls.key and tls.crt, respectively.
+
+//We need to come up with the functions below but for now we can use the assumed defaults
+func certfile_path() string {
+	return "/tmp/k8s-webhook-server/serving-certs/tsl.crt"
+}
+
+func keyfile_path() string {
+	return "/tmp/k8s-webhook-server/serving-certs/tsl.key"
+}
+
+func updateValidationWebhook(vWebhook *v1.ValidatingWebhookConfiguration, caCert []byte) bool {
+	updated := false
+	for idx, webhook := range vWebhook.Webhooks {
+		if bytes.Equal(webhook.ClientConfig.CABundle, caCert) {
+			continue
+		}
+		updated = true
+		webhook.ClientConfig.CABundle = caCert
+		vWebhook.Webhooks[idx] = webhook
+	}
+	return updated
+}
+
+func updateMutatingWebhook(mWebhook *v1.MutatingWebhookConfiguration, caCert []byte) bool {
+	updated := false
+	for idx, webhook := range mWebhook.Webhooks {
+		if bytes.Equal(webhook.ClientConfig.CABundle, caCert) {
+			continue
+		}
+		updated = true
+		webhook.ClientConfig.CABundle = caCert
+		mWebhook.Webhooks[idx] = webhook
+	}
+	return updated
+}
```

Comment:
This two functions area almost idential, but for the different types in their signature. There is probably a better way to do this. Maybe using interfaces.

Reply:
irrelevant now